server/server_setup.py

```diff
@@ -0,0 +1,241 @@
+    """
+    Fetch the server image using singularity.
+
+    :param `server_dir`: location of all server related files.
+    :param `image_name`: name of the image when fetched.
+    """
+    if not os.path.exists(server_dir):
+        LOG.info("Creating merlin server directory.")
+        os.mkdir(server_dir)
+
+    image_path = os.path.join(server_dir, image_name)
+
+    if os.path.exists(image_path):
+        LOG.info(f"{image_path} already exists.")
+        return False
+
+    LOG.info("Fetching redis image from docker://redis.")
+    subprocess.run(["singularity", "pull", image_path, "docker://redis"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    LOG.info("Copying default redis configuration file.")
+    try:
+        shutil.copy(os.path.dirname(os.path.abspath(__file__)) + "/" + CONFIG_FILE, server_dir)
+    except OSError:
+        LOG.error(f"Destination location {server_dir} is not writable.")
+        return False
+    return True
+
+
+def start_server(server_dir: str = SERVER_DIR, image_name: str = IMAGE_NAME):
+    """
+    Start a merlin server container using singularity.
+
+    :param `server_dir`: location of all server related files.
+    :param `image_name`: name of the image when fetched.
+    """
+    current_status = get_server_status(server_dir=server_dir, image_name=image_name)
+
+    if current_status == ServerStatus.NOT_INITALIZED or current_status == ServerStatus.MISSING_CONTAINER:
+        LOG.info("Merlin server has not been initialized. Please run 'merlin server init' first.")
+        return False
+
+    if current_status == ServerStatus.RUNNING:
+        LOG.info("Merlin server already running.")
+        LOG.info("Stop current server with 'merlin server stop' before attempting to start a new server.")
+        return False
+
+    process = subprocess.Popen(
+        ["singularity", "run", os.path.join(server_dir, image_name), os.path.join(server_dir, CONFIG_FILE)],
+        start_new_session=True,
+        close_fds=True,
+        stdout=subprocess.DEVNULL,
+    )
+
+    with open(server_dir + PID_FILE, "w+") as f:
+        f.write(str(process.pid))
+
+    time.sleep(1)
+
+    if get_server_status(server_dir=server_dir, image_name=image_name) != ServerStatus.RUNNING:
+        LOG.error("Unable to start merlin server.")
+        return False
+
+    LOG.info(f"Server started with PID {str(process.pid)}")
+    return True
+
+
+def stop_server(server_dir: str = SERVER_DIR, image_name: str = IMAGE_NAME):
+    """
+    Stop running merlin server containers.
+
+    :param `server_dir`: location of all server related files.
+    :param `image_name`: name of the image when fetched.
+    """
+    if get_server_status(server_dir=server_dir, image_name=image_name) != ServerStatus.RUNNING:
+        LOG.info("There is no instance of merlin server running.")
+        LOG.info("Start a merlin server first with 'merlin server start'")
+        return False
+
```

Comment:
Check for a non-empty stop_command and kill if not present.

Reply:
Checks have been added to ensure that necessary command values are defined in the config file.